### ISSUE TITLE
CORS problem.. Reverse proxy and authentication

### ISSUE THREAD

**Issue by aronmgv** [wukko/cobalt#356]
Hello,

I would like to run this service behind reverse proxy and only expose `cobalt-web` for the access and secure it with authentication such as authelia.

How I would imagine my `docker-compose.yml` to look like:
```yml
services:
    cobalt-api:
        image: ghcr.io/wukko/cobalt:7
        restart: unless-stopped
        container_name: cobalt-api
        init: true
        ports:
            - 9000:9000/tcp
        environment:
            - apiURL=http://cobalt-api/
            - apiName=cobalt-api

    cobalt-web:
        image: ghcr.io/wukko/cobalt:7
        restart: unless-stopped
        container_name: cobalt-web
        init: true
        ports:
            - 9001:9001/tcp
        environment:
            - webURL=http://cobalt-web/
            - apiURL=http://cobalt-api/
```

Reverse proxy configuration pointing `cobalt.example.com` to `localhost:9001`.

But the web is not contacting `cobalt-api` service but only redirecting the connection:
![image](https://github.com/wukko/cobalt/assets/6546231/df12ec26-8d90-4e69-ba95-4bf52feba394)

Is it possible to have such configuration where there is no need to expose `cobalt-api`?

Then I tried exposing both:
```yml
version: '3.7'

services:
    cobalt-api:
        image: ghcr.io/wukko/cobalt:7
        restart: unless-stopped
        container_name: cobalt-api
        init: true
        ports:
            - 9000:9000/tcp
        environment:
            - cors=0
            - webURL=https://cobalt.example.com/
            # replace apiURL with your instance's target url in same format
            - apiURL=https://cobalt-api.example.com/
            # replace apiName with your instance's distinctive name
            - apiName=cobalt-api.example.com

    cobalt-web:
        image: ghcr.io/wukko/cobalt:7
        restart: unless-stopped
        container_name: cobalt-web
        init: true
        ports:
            - 9001:9001/tcp
        environment:
            # replace webURL with your instance's target url in same format
            - webURL=https://cobalt.example.com/
            # replace apiURL with preferred api instance url
            - apiURL=https://cobalt-api.example.com/
```
This leads to following CORS error. I am aware I can disable it at reverse proxy level but dont like this option and when I put authetication it starts failing at 401.. since authentication is not forwarded from `cobalt-web` to `cobalt-api`.. 

`Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://cobalt-api.example.com/api/json. (Reason: CORS request did not succeed). Status code: (null).`

So my question is how am I supposed to configure this setup in a way where only `cobalt-web` is on the public domain (behind authentication). If not possible how to make it work without altering with security on my reverse proxy setup. Thanks

**Comment by dumbmoron** [comment]
hi,
> Is it possible to have such configuration where there is no need to expose cobalt-api?

you will always need to expose it in some way, as the "web" part is just a frontend that does not handle any downloading logic. one way of exposing it while keeping `cors=0` could be to put it on the same (sub)domain as cobalt-web, but on a subpath.

for example, suppose that you want to use the subdomain `cobalt.example.com` for your instance. You could do:
```yaml
version: '3.7'

services:
    cobalt-api:
        image: ghcr.io/wukko/cobalt:7
        restart: unless-stopped
        container_name: cobalt-api
        init: true
        ports:
            - 127.0.0.1:9000:9000/tcp
        environment:
            - cors=0
            - webURL=https://cobalt.example.com/
            - apiURL=https://cobalt.example.com/
            - apiName=cobalt-api

    cobalt-web:
        image: ghcr.io/wukko/cobalt:7
        restart: unless-stopped
        container_name: cobalt-web
        init: true
        ports:
            - 127.0.0.1:9001:9001/tcp
        environment:
            - webURL=https://cobalt.example.com/
            - apiURL=https://cobalt.example.com/
```

and then configure your webserver to proxy the requests accordingly.
i don't know what web server you use, but here is a minimal example of how such configuration could look in nginx (sans additional SSL configuration):
```nginx
# /etc/nginx/conf.d/cobalt.conf
upstream cobalt-api {
    server 127.0.0.1:9000  max_fails=0;
}

upstream cobalt-web {
    server 127.0.0.1:9001  max_fails=0;
}

server {
	listen 80;
	server_name cobalt.example.com;

	location / {
		proxy_pass http://cobalt-web;
	}

        location /api {
		proxy_pass http://cobalt-api/api;
        }
}
```

**Comment by DrWarpMan** [comment]
@dumbmoron I am using almost the exact setup of reverse proxy that you provided, but via [Traefik](https://github.com/traefik/traefik). It took me (literally) the entire day to figure that out.

Can you confirm, that there is no straightforward way to setup authentication in front of cobalt, while having frontend and backend on two different domains, because of the CORS issue?  (e.g.: _cobalt.domain.tld_ & _cobalt-api.domain.tld_)

**Comment by dumbmoron** [comment]
> Can you confirm, that there is no straightforward way to setup authentication in front of cobalt, while having frontend and backend on two different domains, because of the CORS issue? (e.g.: _cobalt.domain.tld_ & _cobalt-api.domain.tld_)

The problem is not so much CORS in this case (that can be taken care of by setting the webURL env for the API if `cors=0` is set, or setting `cors=1`) as that you will need to make sure that when you authenticate against `cobalt.domain.tld` with whatever authentication mechanism you put there, that this session will also apply to `cobalt-api.domain.tld` - for example, when using cookies, your authentication mechanism would need to set a cookie for `.domain.tld` (which basically means it applies for `*.domain.tld`) and not `cobalt.domain.tld`, because then you would obviously not be authenticated against the API subdomain (and you have no real way of authenticating against it anyways).

**Comment by dumbmoron** [comment]
actually, reviewing the source code, yes, you are right in that you are unable to set the CORS origin due to how cobalt determines whether it should run in API or web mode -- this is most likely an oversight, and so i will try to think of a way to fix this

**Comment by aronmgv** [comment]
Thanks both of you. Managed to successfully configure it as suggested.

**Comment by wukko** [comment]
> you are right in that you are unable to set the CORS origin due to how cobalt determines whether it should run in API or web mode

fixed in 7.10.3, webURL was replaced with `CORS_URL` just like in wip 8.0 :D
